node/src/p2p/swarm_manager.rs

```diff
@@ -0,0 +1,703 @@
+            }
+            _ => trace!("Unhandled Kademlia event"),
+        }
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    fn on_ping_event(&mut self, ev: ping::Event) {
+        self.peer_tracker.on_ping_event(&ev);
+
+        match ev.result {
+            Ok(dur) => debug!(
+                "Ping success: peer: {}, connection_id: {}, time: {:?}",
+                ev.peer, ev.connection, dur
+            ),
+            Err(e) => {
+                debug!(
+                    "Ping failure: peer: {}, connection_id: {}, error: {}",
+                    &ev.peer, &ev.connection, e
+                );
+                self.swarm.close_connection(ev.connection);
+            }
+        }
+    }
+
+    pub(crate) async fn stop(&mut self) {
+        self.swarm
+            .behaviour_mut()
+            .connection_control
+            .set_stopping(true);
+
+        for listener in self.listeners.drain(..) {
+            self.swarm.remove_listener(listener);
+        }
+
+        for (_, connection_id) in self.peer_tracker.all_connections() {
+            self.swarm.close_connection(connection_id);
+        }
+
+        // Waiting until all established connections are closed.
+        while self
+            .swarm
+            .network_info()
+            .connection_counters()
+            .num_established()
+            > 0
+        {
+            match self.swarm.select_next_some().await {
+                // We may receive this if connection was established just before we trigger stop.
+                SwarmEvent::ConnectionEstablished { connection_id, .. } => {
+                    // We immediately close the connection in this case.
+                    self.swarm.close_connection(connection_id);
+                }
+                SwarmEvent::ConnectionClosed {
+                    peer_id,
+                    connection_id,
+                    ..
+                } => {
+                    // This will generate the PeerDisconnected events.
+                    self.on_peer_disconnected(&peer_id, connection_id);
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+fn init_kademlia(
+    network_id: &str,
+    keypair: &Keypair,
+    listen_on: &[Multiaddr],
+) -> Result<kad::Behaviour<kad::store::MemoryStore>> {
+    let local_peer_id = PeerId::from(keypair.public());
+    let store = kad::store::MemoryStore::new(local_peer_id);
+
+    let protocol_id = celestia_protocol_id(network_id, "/kad/1.0.0");
+    let config = kad::Config::new(protocol_id);
+
+    let mut kademlia = kad::Behaviour::with_config(local_peer_id, store, config);
+
+    if !listen_on.is_empty() {
+        kademlia.set_mode(Some(kad::Mode::Server));
+    }
+
+    Ok(kademlia)
+}
+
+/// Converts a topic to `RecordKey`.
+///
+/// This is the equivalent of `nsToCid` that is used in [`RoutingDiscovery.FindPeers`][1]
+/// of go-libp2p which later on is unwraped to the internal hash in [`IpfsDHT.FindProvidersAsync`][2].
```

Comment:
```suggestion
/// of go-libp2p which later on is unwraped to internal hash in [`IpfsDHT.FindProvidersAsync`][2].
```
my article senses are tingling